### DISCUSSION TITLE
Replace "On hand" with multiple statusses

### DISCUSSION THREAD

**Discussion by ehuelsmann** [ledgersmb/LedgerSMB#5818] [Ideas]
On hand numbers are off with respect too the inventory tracking that's part of the code code, regularly.
We need to define on hand as the number of items received minus the number of items shipped.l? Or rather the number of items bought minus sold? (The latter equating to the number reported in the inventory account...)

But what about figures like number of items bought minus number of items received (incoming, in transit) or the number of items shipped minus those sold (in transit, outgoing).
And there's ordered but not received, ordered but not shipped.

Each of these has value in a specific context in the company.

**Comment by ehuelsmann** [comment]
Looking at the picture below, On Hand can be "In Stock" or "In Stock minus Ordered-but-not-shipped".

![stock-vs-inventory](https://user-images.githubusercontent.com/2326559/95009754-b8908380-0624-11eb-80f5-b4a10eb6075e.png)

**Comment by ehuelsmann** [comment]
@tonka128, can you relate to this diagram? Also, if this diagram were to be leading, it would mean users of our application need to use the shipping module to get the right numbers in their OnHand figure (whichever definition we'll give On Hand)...

Do you currently use the Shipping module? If not, would you consider being required to use it to get correct "OnHand" numbers, a problem?

**Comment by ehuelsmann** [comment]
The statusses I'm thinking about now are:

1. (Back)order / In Stock (Available) / In Stock (Reserved by order)
2. Inventory

The numbers "In Stock" and "Inventory" are obviously linked, but need not be equal. In order to understand the differences between the two, I think there ought to be some tracking between what's received and what's purchased as well as between what's sold and what's shipped.

**Comment by tonka128** [comment]
Thats a lot more involved then our setup at the moment, we are not using the shipping module at all (is that what causes the break?) - so our inventory levels are just whats been purchased minus whats been sold... we only enter a purchase once its been physically received and we only convert a SO to an SI as it goes out the door. leading my expectation to be the number of an item in the system is whats available to sell and any point.

**Comment by ehuelsmann** [comment]
Ok. Thanks for confirming this is bigger than you are doing now. I think your business is a nice model of our target audience. That being the case, I think we need to cater to it and come up with a way to add to stock and inventory at the same time.

**Comment by ehuelsmann** [comment]
No, the fact that you skip shipping isn't causing the breakage. Rather, there are onhand updates in code for orders as well as invoices. The various code paths need to be researched for duplication.
That or we need to create other means to track whether parts have already been added to inventory and stock.

**Comment by ehuelsmann** [comment]
Further thinking about the matter, I think that this table below describes the relation between "in stock" and "in inventory".

Inventory vs Stock: Impact of shipping and invoicing

  |                      | to receive <br /> ordered   | in stock         | in stock <br />ordered/reserved             | shipped |
  | :--: | :--: | :--: | :--: | :--: |
  | to be invoiced <br /> to be purchased         | - / -      | - / +    | - / +*              | - / -   |
  | inventory <br /> purchased, to be sold | + / -      | + / +    | + / +*              | + / -   |
  | invoiced <br /> sold                | - / -      | - / +    | - / +*              | - / -   |

* In stock, but reserved to order; do we need an additional state _not available_?

The conclusion must be that only in 6 out of 12 states, inventory and stock levels are expected to be equal, however, the remaining 6 states are expected to be relatively short-lived (when shipping is used): they indicate situations where purchase/sale and shipping are out of sync due to timing.

Given @tonka128's reaction above, we'll probably need a "knob" to select whether to report "on hand" based on stock levels or on inventory levels. Additionally, when companies enable the use of shipping, they will want to have tables like the table above available to understand any differences in numbers.

**Comment by ehuelsmann** [comment]
Side note: It would be interesting to figure out what stock counts should do to inventory numbers, if we decide to go with the table above.

**Comment by sbts** [comment]
Just to clarify. 
I believe in the past (and possibly currently) an order removed items from on hand.

I can't remember further details on that, eg: how to release the held goods without converting the order to an invoice

**Comment by ehuelsmann** [comment]
@sbts I think you're correct about the order removing the goods. However, I think that the invoice does too, except when ... Except that I'm not sure what the exceptional cases are... And given the fact that stock and inventory are usually coinciding, it was hard to see why the different numbers wouldn't match. Thinking about this a bit longer, I now understand that stock (stuff in the warehouse) and inventory (stuff you have paid for and hence own) are not exactly the same thing -- but so similar that they can be easily confused, which I think is what happened in our code: to keep things simple, they were lumped into one and the same.
